## Куликов Василий
## ДЗ 1 и ДЗ 2

In [5]:
apostrophe_dict = {
"ain't": "am not / are not",
"aren't": "are not / am not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is",
"i'd": "I had / I would",
"i'd've": "I would have",
"i'll": "I shall / I will",
"i'll've": "I shall have / I will have",
"i'm": "I am",
"i've": "I have",
"isn't": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}



short_word_dict = {
"121": "one to one",
"a/s/l": "age, sex, location",
"adn": "any day now",
"afaik": "as far as I know",
"afk": "away from keyboard",
"aight": "alright",
"alol": "actually laughing out loud",
"b4": "before",
"b4n": "bye for now",
"bak": "back at the keyboard",
"bf": "boyfriend",
"bff": "best friends forever",
"bfn": "bye for now",
"bg": "big grin",
"bta": "but then again",
"btw": "by the way",
"cid": "crying in disgrace",
"cnp": "continued in my next post",
"cp": "chat post",
"cu": "see you",
"cul": "see you later",
"cul8r": "see you later",
"cya": "bye",
"cyo": "see you online",
"dbau": "doing business as usual",
"fud": "fear, uncertainty, and doubt",
"fwiw": "for what it's worth",
"fyi": "for your information",
"g": "grin",
"g2g": "got to go",
"ga": "go ahead",
"gal": "get a life",
"gf": "girlfriend",
"gfn": "gone for now",
"gmbo": "giggling my butt off",
"gmta": "great minds think alike",
"h8": "hate",
"hagn": "have a good night",
"hdop": "help delete online predators",
"hhis": "hanging head in shame",
"iac": "in any case",
"ianal": "I am not a lawyer",
"ic": "I see",
"idk": "I don't know",
"imao": "in my arrogant opinion",
"imnsho": "in my not so humble opinion",
"imo": "in my opinion",
"iow": "in other words",
"ipn": "I’m posting naked",
"irl": "in real life",
"jk": "just kidding",
"l8r": "later",
"ld": "later, dude",
"ldr": "long distance relationship",
"llta": "lots and lots of thunderous applause",
"lmao": "laugh my ass off",
"lmirl": "let's meet in real life",
"lol": "laugh out loud",
"ltr": "longterm relationship",
"lulab": "love you like a brother",
"lulas": "love you like a sister",
"luv": "love",
"m/f": "male or female",
"m8": "mate",
"milf": "mother I would like to fuck",
"oll": "online love",
"omg": "oh my god",
"otoh": "on the other hand",
"pir": "parent in room",
"ppl": "people",
"r": "are",
"rofl": "roll on the floor laughing",
"rpg": "role playing games",
"ru": "are you",
"shid": "slaps head in disgust",
"somy": "sick of me yet",
"sot": "short of time",
"thanx": "thanks",
"thx": "thanks",
"ttyl": "talk to you later",
"u": "you",
"ur": "you are",
"uw": "you’re welcome",
"wb": "welcome back",
"wfm": "works for me",
"wibni": "wouldn't it be nice if",
"wtf": "what the fuck",
"wtg": "way to go",
"wtgp": "want to go private",
"ym": "young man",
"gr8": "great"
}


emoticon_dict = {
":)": "happy",
":‑)": "happy",
":-]": "happy",
":-3": "happy",
":->": "happy",
"8-)": "happy",
":-}": "happy",
":o)": "happy",
":c)": "happy",
":^)": "happy",
"=]": "happy",
"=)": "happy",
"<3": "happy",
":-(": "sad",
":(": "sad",
":c": "sad",
":<": "sad",
":[": "sad",
">:[": "sad",
":{": "sad",
">:(": "sad",
":-c": "sad",
":-< ": "sad",
":-[": "sad",
":-||": "sad"
}

In [6]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import nltk
import warnings 
warnings.filterwarnings("ignore", category=DeprecationWarning)
import os
import html

In [7]:
'''nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')'''

"nltk.download('punkt')\nnltk.download('stopwords')\nnltk.download('wordnet')"

In [8]:
train_df = pd.read_csv('train_tweets.csv')
train_df.head()

id  label                                              tweet
0   1      0   @user when a father is dysfunctional and is s...
1   2      0  @user @user thanks for #lyft credit i can't us...
2   3      0                                bihday your majesty
3   4      0  #model   i love u take with u all the time in ...
4   5      0             factsguide: society now    #motivation

In [9]:
test_df = pd.read_csv('test_tweets.csv')
test_df.head()

id                                              tweet
0  31963  #studiolife #aislife #requires #passion #dedic...
1  31964   @user #white #supremacists want everyone to s...
2  31965  safe ways to heal your #acne!!    #altwaystohe...
3  31966  is the hp and the cursed child book up for res...
4  31967    3rd #bihday to my amazing, hilarious #nephew...

In [10]:
combine_df = train_df.append(test_df, ignore_index = True, sort = False)
combine_df.head()

id  label                                              tweet
0   1    0.0   @user when a father is dysfunctional and is s...
1   2    0.0  @user @user thanks for #lyft credit i can't us...
2   3    0.0                                bihday your majesty
3   4    0.0  #model   i love u take with u all the time in ...
4   5    0.0             factsguide: society now    #motivation

In [11]:
print(combine_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49159 entries, 0 to 49158
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      49159 non-null  int64  
 1   label   31962 non-null  float64
 2   tweet   49159 non-null  object 
dtypes: float64(1), int64(1), object(1)
memory usage: 1.1+ MB
None


### HW Lesson1

1. Заменим html-сущности (к примеру: &lt; &gt; &amp;). "&lt;" заменим на “<” и "&amp;" заменим на “&”)""". Сделаем это с помощью HTMLParser.unescape()

2. Удалим @user из всех твитов с помощью паттерна "@[\w]*". Для этого создадим функцию: 
 - для того, чтобы найти все вхождения паттерна в тексте, необходимо использовать re.findall(pattern, input_txt)
 - для для замены @user на пробел, необходимо использовать re.sub()
при применении функции необходимо использовать np.vectorize(function)


3. Изменим регистр твитов на нижний с помощью .lower()

4. Заменим сокращения с апострофами (пример: ain't, can't) на пробел, используя apostrophe_dict. Для этого необходимо сделать функцию: для каждого слова в тексте проверить (for word in text.split()), если слово есть в словаре apostrophe_dict в качестве ключа (сокращенного слова), то заменить ключ на значение (полную версию слова).

5. Заменим сокращения на их полные формы, используя short_word_dict. Для этого воспользуемся функцией, используемой в предыдущем пункте.

6. Заменим эмотиконы (пример: ":)" = "happy") на пробелы, используя emoticon_dict. Для этого воспользуемся функцией, используемой в предыдущем пункте.

7. Заменим пунктуацию на пробелы, используя re.sub() и паттерн r'[^\w\s]'

8. Заменим спец. символы на пробелы, используя re.sub() и паттерн r'[^a-zA-Z0-9]'

9. Заменим числа на пробелы, используя re.sub() и паттерн r'[^a-zA-Z]'

10. Удалим из текста слова длиной в 1 символ, используя ' '.join([w for w in x.split() if len(w)>1])

11. Поделим твиты на токены с помощью nltk.tokenize.word_tokenize, создав новый столбец 'tweet_token'.

12. Удалим стоп-слова из токенов, используя nltk.corpus.stopwords. Создадим столбец 'tweet_token_filtered' без стоп-слов.

13. Применим стемминг к токенам с помощью nltk.stem.PorterStemmer. Создадим столбец 'tweet_stemmed' после применения стемминга.

14. Применим лемматизацию к токенам с помощью nltk.stem.wordnet.WordNetLemmatizer. Создадим столбец 'tweet_lemmatized' после применения лемматизации.

15. Сохраним результат предобработки в pickle-файл.

In [12]:
### 1. Заменим html-сущности (к примеру: &lt; &gt; &amp;). "&lt;" заменим на “<” и "&amp;" заменим на “&”)""".
### Сделаем это с помощью HTMLParser.unescape()
def replace_special_symbol(text):
    return  html.unescape(text)

### 2. Удалим @user из всех твитов с помощью паттерна "@[\w]*". Для этого создадим функцию: 
### - для того, чтобы найти все вхождения паттерна в тексте, необходимо использовать re.findall(pattern, input_txt)
### - для для замены @user на пробел, необходимо использовать re.sub()
### при применении функции необходимо использовать np.vectorize(function)
def delete_users_name(text):
    return re.sub('@[\w]*', ' ', text)

### 3. Изменим регистр твитов на нижний с помощью .lower()
def lower_tweets(text):
    return str(text).lower()

### 4. Заменим сокращения с апострофами (пример: ain't, can't) на пробел, используя apostrophe_dict. 
### Для этого необходимо сделать функцию: для каждого слова в тексте проверить (for word in text.split()),
### если слово есть в словаре apostrophe_dict в качестве ключа (сокращенного слова), то заменить ключ на значение (полную версию слова).

def  replace_words_in_dictonary(text, dictionary_words):     
    return ' '.join(dictionary_words.get(word, word) for word in text.split())

def replace_cutwords(text):
    global apostrophe_dict    
    return replace_words_in_dictonary(text, apostrophe_dict)

### 5. Заменим сокращения на их полные формы, используя short_word_dict. 
### Для этого воспользуемся функцией, используемой в предыдущем пункте.
def replace_short_words(text):
    global short_word_dict    
    return replace_words_in_dictonary(text, short_word_dict)

### 6. Заменим эмотиконы (пример: ":)" = "happy") на пробелы, используя emoticon_dict. 
### Для этого воспользуемся функцией, используемой в предыдущем пункте.
def replace_emoticon(text):
    global emoticon_dict
    return replace_words_in_dictonary(str(text), emoticon_dict)

### 7. Заменим пунктуацию на пробелы, используя re.sub() и паттерн r'[^\w\s]'
def replace_punctuation_mark(text):
    return re.sub(r'[^\w\s]', ' ', text)

### 8. Заменим спец. символы на пробелы, используя re.sub() и паттерн r'[^a-zA-Z0-9]'
def replace_other_symbol(text):
    return re.sub(r'[^a-zA-Z0-9]', ' ', text)

### 9. Заменим числа на пробелы, используя re.sub() и паттерн r'[^a-zA-Z]'
def replace_numbers(text):
    return re.sub(r'[^a-zA-Z]', ' ', text)

### 10. Удалим из текста слова длиной в 1 символ, используя ' '.join([w for w in x.split() if len(w)>1])
def delete_lonely_char(text):
    return ' '.join([word for word in text.split() if len(word)>1])

In [13]:
def preprocess(text):
    func = [replace_special_symbol, delete_users_name, delete_users_name, lower_tweets, replace_cutwords,
            replace_short_words, replace_emoticon, replace_punctuation_mark, replace_other_symbol, replace_numbers, delete_lonely_char]
    result = text
    for f in func:
        result = f(result)
    return result

In [14]:
combine_df['clean_tweet'] = combine_df['tweet'].apply(lambda x: preprocess(x))

In [15]:
combine_df['clean_tweet'].head(10)

0    when father is dysfunctional and is so selfish...
1    thanks for lyft credit cannot use cause they d...
2                                  bihday your majesty
3      model love you take with you all the time in ur
4                    factsguide society now motivation
5    huge fan fare and big talking before they leav...
6                               camping tomorrow danny
7    the next school year is the year for exams can...
8    we won love the land allin cavs champions clev...
9                   welcome here am it has it is so gr
Name: clean_tweet, dtype: object

Поделим твиты на токены с помощью nltk.tokenize.word_tokenize, создав новый столбец 'tweet_token'.

In [16]:
combine_df['tweet_token'] = combine_df.clean_tweet.apply(lambda x: nltk.word_tokenize(x))
combine_df['tweet_token'].head(2)

0    [when, father, is, dysfunctional, and, is, so,...
1    [thanks, for, lyft, credit, can, not, use, cau...
Name: tweet_token, dtype: object

Удалим стоп-слова из токенов, используя nltk.corpus.stopwords. Создадим столбец 'tweet_token_filtered' без стоп-слов.

In [17]:
stop_words = nltk.corpus.stopwords.words('english')
combine_df['tweet_token_filtered' ] = combine_df.tweet_token.apply(lambda x: [token for token in x if token not in stop_words])

Применим стемминг к токенам с помощью nltk.stem.PorterStemmer. Создадим столбец 'tweet_stemmed' после применения стемминга.

In [18]:
stemmer = nltk.stem.PorterStemmer()
combine_df['tweet_stemmed'] = combine_df.tweet_token_filtered.apply(lambda x: [stemmer.stem(token) for token in x])

Применим лемматизацию к токенам с помощью nltk.stem.wordnet.WordNetLemmatizer. Создадим столбец 'tweet_lemmatized' после применения лемматизации.

In [19]:
lemmatizer = nltk.stem.wordnet.WordNetLemmatizer()
combine_df['tweet_lemmatized'] = combine_df.tweet_token_filtered.apply(lambda x: [lemmatizer.lemmatize(token) for token in x])

Сохраним результат предобработки в pickle-файл.

In [19]:
combine_df.to_pickle('./tweet_data.pkl')

### HW Lesson2

Тема “Создание признакового пространства”

Продолжим обработку данных с Твиттера. 

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
Исключим стоп-слова с помощью stop_words='english'.
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().
3. Натренируем gensim.models.Word2Vec модель на наших данных.
Тренировать будем на токенизированных твитах combine_df['tweet_token']
Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).


Чат-бот
Докрутить чат-бота (обучить лучший ft, w2v, взвесить idf). Модели сдампить и сложить в облако, ноутбук можно через гитхаб. Ответы: https://yadi.sk/d/SLgpFrmrBubeYg

In [3]:
import warnings
warnings.filterwarnings("ignore")

In [20]:
# 1
# Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). 
# Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно. 
# Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df. 
# Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000. 
# Исключим стоп-слова с помощью stop_words='english'. 
# Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [28]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

count_vectorizer = CountVectorizer(max_df=0.9, max_features=1000, stop_words='english')

documents = train_df['tweet'].values

# Создаем the Bag-of-Words модель
bag_of_words = count_vectorizer.fit_transform(documents)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
bag_of_words_count_vec_train = pd.DataFrame(bag_of_words.toarray(), columns = feature_names)

documents = test_df['tweet'].values

# Создаем the Bag-of-Words модель
bag_of_words = count_vectorizer.fit_transform(documents)

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
bag_of_words_count_vec_test = pd.DataFrame(bag_of_words.toarray(), columns = feature_names)

In [29]:
print(bag_of_words_count_vec_train)
print(bag_of_words_count_vec_test)

       00  06  10  100  11  12  13  14  15  16  ...  µð  ¹ð  ºï  ºð  ¼ï  ¼ð  \
0       0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
1       0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
2       0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
3       0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
4       0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
...    ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..   
31957   0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
31958   0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
31959   0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
31960   0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   
31961   0   0   0    0   0   0   0   0   0   0  ...   0   0   0   0   0   0   

       ½ï  ½ð  ¾ð  ó¾  
0       0   0   0   0  
1  

In [24]:
# 2
# Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
# Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
# Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
# Исключим стоп-слова с помощью stop_words='english'.
# Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [30]:
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd

document = train_df['tweet'].values
tfidf_vectorizer = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')
values = tfidf_vectorizer.fit_transform(document)

# Show the Model as a pandas DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
bag_of_words_tfidf_vec_train = pd.DataFrame(values.toarray(), columns = feature_names)

document = test_df['tweet'].values
tfidf_vectorizer = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')
values = tfidf_vectorizer.fit_transform(document)

# Show the Model as a pandas DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
bag_of_words_tfidf_vec_test = pd.DataFrame(values.toarray(), columns = feature_names)

In [31]:
print(bag_of_words_tfidf_vec_train)
print(bag_of_words_tfidf_vec_test)

        00   06   10  100   11   12   13   14   15   16  ...   µð   ¹ð   ºï  \
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
31957  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
31958  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
31959  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
31960  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
31961  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

        ºð   ¼ï   ¼ð   ½ï   ½ð   ¾ð   ó¾  
0      0

In [ ]:
# 3
# Натренируем gensim.models.Word2Vec модель на наших данных.
# Тренировать будем на токенизированных твитах combine_df['tweet_token']
# Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
# Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [32]:
from gensim.test.utils import common_texts, get_tmpfile
from gensim.models import Word2Vec

model = Word2Vec(combine_df['tweet_token'], size=200, window=5, min_count=2, 
                 sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)
# model.save("word2vec.model")

In [36]:
model.train(combine_df['tweet_token'], total_examples=31962, epochs=20)

(9142668, 11726520)

In [38]:
# 4
# Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. 
# Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. 
# То же самое попробуем со словом "trump".

In [40]:
model.most_similar(positive=['dinner'])

[('bihdaydinner', 0.5549252033233643),
 ('spaghetti', 0.5500830411911011),
 ('bolognese', 0.5482354164123535),
 ('tacotuesday', 0.5413339138031006),
 ('lastnight', 0.5380694270133972),
 ('cookout', 0.5301350355148315),
 ('shawarma', 0.5248984098434448),
 ('waterloo', 0.5223593711853027),
 ('burritos', 0.5204410552978516),
 ('hamburger', 0.5187911987304688)]

In [41]:
model.most_similar(positive=['trump'])

[('donald', 0.541990339756012),
 ('suppoer', 0.527134358882904),
 ('fuhered', 0.526708722114563),
 ('dumptrump', 0.5119016170501709),
 ('delegaterevolt', 0.5088772177696228),
 ('unfit', 0.4986211657524109),
 ('impeachment', 0.49253666400909424),
 ('conman', 0.4921892285346985),
 ('unfavorability', 0.49216485023498535),
 ('donaldtrump', 0.49018439650535583)]

In [42]:
# 5
# Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для 
# данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
# Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [43]:
model.wv['food']

array([ 0.31965262, -0.08884489, -0.33429065,  0.2842012 , -0.59109503,
        0.3882808 , -0.20148955,  0.15986131,  0.517856  , -0.38417476,
       -0.4064829 ,  0.23117447, -0.18508486,  0.4922358 ,  0.50463533,
        0.04868267,  0.24252021, -0.31308693, -0.7200623 ,  0.5773206 ,
        0.48437276, -0.40226564, -0.81536955,  0.44971016, -0.10897569,
        0.13141234, -0.00708144, -0.55593014,  0.02323177, -0.3470028 ,
        0.49584618, -0.2772987 ,  0.44329467,  0.30354968, -0.04934626,
        0.85290575, -0.6327217 ,  0.795553  ,  0.00525001,  0.311322  ,
        0.091803  , -0.08621893,  0.20573023, -1.0911386 ,  0.5112627 ,
        0.19114442, -0.08674183, -0.648637  ,  0.25081903,  0.25275153,
       -0.50833017,  0.23165481,  0.7260811 , -0.3359345 , -0.20373632,
       -0.74248195, -0.26962474, -0.26752004, -0.24287288,  0.05036914,
        0.19714087, -0.10741026,  0.28473917,  0.17744792, -0.4061352 ,
       -0.26164988,  0.5134973 ,  0.40269408, -0.2623409 , -0.79